packages/core/src/components/forms/numericStepper.tsx

```diff
@@ -0,0 +1,318 @@
+ */
+
+import * as classNames from "classnames";
+import * as PureRender from "pure-render-decorator";
+import * as React from "react";
+
+import * as Classes from "../../common/classes";
+import * as Errors from "../../common/errors";
+import * as Keys from "../../common/keys";
+import { HTMLInputProps, IIntentProps, IProps, removeNonHTMLProps } from "../../common/props";
+
+import { Button } from "../button/buttons";
+import { InputGroup } from "./inputGroup";
+
+export type ButtonPosition = "none" | "left" | "right" | "split";
+
+export interface INumericStepperProps extends IIntentProps, IProps {
+
+    /**
+     * The button configuration with respect to the input field.
+     * @default "right"
+     */
+    buttonPosition?: ButtonPosition;
+
+    /**
+     * Whether the input is non-interactive.
+     * @default false
+     */
+    disabled?: boolean;
+
+    /** Ref handler that receives HTML `<input>` element backing this component. */
+    inputRef?: (ref: HTMLInputElement) => any;
+
+    /** Name of icon (the part after `pt-icon-`) to render on left side of input. */
+    leftIconName?: string;
+
+    /** Placeholder text in the absence of any value. */
+    placeholder?: string;
+
+    /**
+     * Increment between successive values when `shift` is held.
+     * @default 10
+     */
+    majorStepSize?: number;
+
+    /** Maximum value of the input. */
+    max?: number;
+
+    /** Minimum value of the input. */
+    min?: number;
+
+    /**
+     * Increment between successive values when `alt` is held.
+     * @default 0.1
+     */
+    minorStepSize?: number;
+
+    /**
+     * Increment between successive values when no modifier keys are held.
+     * @default 1
+     */
+    stepSize?: number;
+
+    /**
+     * Value to display in the input field
+     * @default ""
+     */
+    value?: number | string;
+
+    /** Callback invoked when the value changes. */
+    onChange?(value: string): void;
+}
+
+export interface INumericStepperState {
+    shouldSelectAfterUpdate?: boolean;
+    value?: string;
+}
+
+@PureRender
+export class NumericStepper extends React.Component<HTMLInputProps & INumericStepperProps, INumericStepperState> {
+    public static displayName = "Blueprint.NumericStepper";
+
+    public static defaultProps: INumericStepperProps = {
+        buttonPosition: "right",
+        majorStepSize: 10,
+        minorStepSize: 0.1,
+        stepSize: 1,
+        value: NumericStepper.VALUE_EMPTY,
+    };
+
+    private static DECREMENT_KEY = "decrement";
+    private static INCREMENT_KEY = "increment";
+    private static DECREMENT_ICON_NAME = "minus";
+    private static INCREMENT_ICON_NAME = "plus";
+    private static VALUE_EMPTY = "";
+    private static VALUE_ZERO = "0";
+
+    private inputElement: HTMLInputElement;
+
+    public constructor(props?: HTMLInputProps & INumericStepperProps) {
+        super(props);
+        this.validateProps(props);
+        this.state = {
```

Comment:
Does it make sense to just set state in `componentWillReceiveProps`? I do it there again anyway.